### PR TITLE
Added support for live migration of generated always as identity columns for Oracle and YugabyteDB

### PR DESCRIPTION
**Pull request by sanyamsinghal**
- disable them at the start of import data by alter the columns to generated by default
- enable them at the end of import data by altering the columns to generated always

### CODE REVIEW

**Comment by sanyamsinghal** [comment]
Jenkins Pipeline Test Results:
1. With Debezium: https://jenkins.dev.yugabyte.com/job/users/job/yb-voyager-testing/job/yb-voyager-testing-pipeline/1015/

File: yb-voyager/src/tgtdb/target_db_interface.go [hunk 1]
```diff
@@ -43,6 +43,10 @@ type TargetDB interface {
 	RestoreSequences(sequencesLastValue map[string]int64) error
 	GetImportedEventsStatsForTable(tableName string, migrationUuid uuid.UUID) (*EventCounter, error)
 	GetImportedSnapshotRowCountForTable(tableName string) (int64, error)
+	GetGeneratedAlwaysIdentityColumnNamesForTable(table string) ([]string, error)
+	GetGeneratedAlwaysAsIdentityColumnsForTables(tables []string) (map[string][]string, error)
```
Comment (amit-yb) [inline yb-voyager/src/tgtdb/target_db_interface.go:47]: Nit: Following parts of the name should be present in both the names: `As` and `Names`

Comment (amit-yb) [inline yb-voyager/src/tgtdb/target_db_interface.go:47]: Implementation of this method will be the same for all `TargetDB` interface -- iterate through all the `tables` and call the "single table name" version of the same method.

I would suggest it to define this as a separate function in the `importData.go`.

File: yb-voyager/src/tgtdb/yugabytedb.go [hunk 2]
```diff
@@ -1289,3 +1289,115 @@ func (yb *TargetYugabyteDB) GetImportedSnapshotRowCountForTable(tableName string
...
+		if err != nil {
+			log.Errorf("error in querying identity columns for table %s: %v", table, err)
+			return false, fmt.Errorf("error in querying identity columns for table %s: %w", table, err)
+		}
+		defer rows.Close()
+		for rows.Next() {
+			var colName string
+			err = rows.Scan(&colName)
+			if err != nil {
+				log.Errorf("error in scanning identity column name: %v", err)
```
Comment (amit-yb) [inline yb-voyager/src/tgtdb/yugabytedb.go:1315]: No need of repeating error messages in the log if we're already returning an error with the same message. The error will anyway be logged by the top level function e.g. `main` or `utils.ErrExit()`.

File: yb-voyager/src/tgtdb/yugabytedb.go [hunk 3]
```diff
@@ -1289,3 +1289,115 @@ func (yb *TargetYugabyteDB) GetImportedSnapshotRowCountForTable(tableName string
...
+	return nil
+}
+
+func (yb *TargetYugabyteDB) EnableGeneratedAlwaysAsIdentityColumns(tableColumnsMap map[string][]string) error {
+	log.Infof("enabling generated always as identity columns")
+	for table, columns := range tableColumnsMap {
+		qualifiedTblName := yb.qualifyTableName(table)
+		batch := pgx.Batch{}
+		for _, column := range columns {
+			query := fmt.Sprintf(`ALTER TABLE %s ALTER COLUMN %s SET GENERATED ALWAYS`, qualifiedTblName, column)
```
Comment (amit-yb) [inline yb-voyager/src/tgtdb/yugabytedb.go:1379]: Except for the `SET GENERATED ALWAYS` (and log messages) this function is the same as `Disable*` counterpart.

How about having a private method `alterColumns()` function that will be called as:
```
// From Disable*()
err = yb.alterColumns(tableColumnsMap, "SET GENERATED BY DEFAULT")

// From Enable*()
err = yb.alterColumns(tableColumnsMap, "SET GENERATED ALWAYS")
```

File: yb-voyager/src/tgtdb/yugabytedb.go [hunk 4]
```diff
@@ -1289,3 +1289,115 @@ func (yb *TargetYugabyteDB) GetImportedSnapshotRowCountForTable(tableName string
...
+				}
+			}
+			if err := br.Close(); err != nil {
+				log.Errorf("error closing batch of queries to alter default identity columns for table(%s): %v", qualifiedTblName, err)
+				return false, fmt.Errorf("error closing batch of queries to alter default identity columns for table(%s): %w", qualifiedTblName, err)
+			}
+			return false, nil
+		})
+		if err != nil {
+			log.Errorf("error in enabling identity columns for table %s: %v", qualifiedTblName, err)
```
Comment (amit-yb) [inline yb-voyager/src/tgtdb/yugabytedb.go:1398]: Drop these error messages which are already included in the `error`.

File: yb-voyager/src/tgtdb/yugabytedb.go [hunk 5]
```diff
@@ -1289,3 +1289,115 @@ func (yb *TargetYugabyteDB) GetImportedSnapshotRowCountForTable(tableName string
...
+			return fmt.Errorf("error in enabling identity columns for table %s: %w", qualifiedTblName, err)
```
Comment (amit-yb) [inline yb-voyager/src/tgtdb/yugabytedb.go:1399]: Phrase error messages as `action: error`:
```
return fmt.Errorf("enable identity columns for table %s: %w",
    qualifiedTableName, err)
```

File: yb-voyager/src/tgtdb/yugabytedb.go [hunk 6]
```diff
@@ -1289,3 +1289,115 @@ func (yb *TargetYugabyteDB) GetImportedSnapshotRowCountForTable(tableName string
...
+			return fmt.Errorf("error in disabling identity columns for table %s: %w", qualifiedTblName, err)
+		}
+	}
+	return nil
+}
+
+func (yb *TargetYugabyteDB) EnableGeneratedAlwaysAsIdentityColumns(tableColumnsMap map[string][]string) error {
+	log.Infof("enabling generated always as identity columns")
+	for table, columns := range tableColumnsMap {
+		qualifiedTblName := yb.qualifyTableName(table)
```
Comment (amit-yb) [inline yb-voyager/src/tgtdb/yugabytedb.go:1376]: `qualifiedTableName`

File: yb-voyager/src/tgtdb/oracle.go [hunk 7]
```diff
@@ -815,3 +815,95 @@ func (tdb *TargetOracleDB) GetImportedSnapshotRowCountForTable(tableName string)
...
+	return nil
+}
+
+func (tdb *TargetOracleDB) EnableGeneratedAlwaysAsIdentityColumns(tableColumnsMap map[string][]string) error {
+	log.Infof("enabling generated always as identity columns")
+	for table, columns := range tableColumnsMap {
+		qualifiedTblName := tdb.qualifyTableName(table)
+		for _, column := range columns {
+			// LIMIT VALUE - ensures that start it is set to the current value of the sequence
+			query := fmt.Sprintf(`ALTER TABLE %s MODIFY %s GENERATED ALWAYS AS IDENTITY(START WITH LIMIT VALUE)`, qualifiedTblName, column)
```
Comment (amit-yb) [inline yb-voyager/src/tgtdb/oracle.go:894]: Same comments that are given for the `yugabytedb.go` are also applicable here.

File: yb-voyager/cmd/importData.go [hunk 8]
```diff
@@ -245,6 +249,31 @@ func importData(importFileTasks []*ImportFileTask) {
...
+	if err != nil {
+		utils.ErrExit("failed to disable generated always as identity columns: %s", err)
+	}
+	defer func() {
+		err = tdb.EnableGeneratedAlwaysAsIdentityColumns(TableToIdentityColumnNames)
+		if err != nil {
+			utils.ErrExit("failed to enable generated always as identity columns: %s", err)
+		}
+	}()
+
```
Comment (amit-yb) [inline yb-voyager/cmd/importData.go:276]: Factor out in separate functions to have code structure like:
```
disableGeneratedAlwaysAsIdentityColumns(importFileTasks)
defer enableGeneratedAlwaysAsIdentityColumns()
```

File: yb-voyager/src/tgtdb/yugabytedb.go (hunk 2 above)
Comment (sanyamsinghal) [reply yb-voyager/src/tgtdb/yugabytedb.go:1315]: Simplified the error message a bit in these cases.
But here adding more context to it by telling whether the error was during the `querying` or `scanning` part, so will keep that part

File: yb-voyager/src/tgtdb/yugabytedb.go (hunk 3 above)
Comment (sanyamsinghal) [reply yb-voyager/src/tgtdb/yugabytedb.go:1379]: nice idea, done

File: yb-voyager/src/tgtdb/yugabytedb.go [hunk 9]
```diff
@@ -1289,3 +1289,78 @@ func (yb *TargetYugabyteDB) GetImportedSnapshotRowCountForTable(tableName string
...
+			batch.Queue(query)
+		}
+
+		err := yb.connPool.WithConn(func(conn *pgx.Conn) (bool, error) {
+			br := conn.SendBatch(context.Background(), &batch)
+			for i := 0; i < batch.Len(); i++ {
+				_, err := br.Exec()
+				if err != nil {
+					log.Errorf("executing query to alter columns for table(%s): %v", qualifiedTableName, err)
+					return false, fmt.Errorf("executing query to alter columns for table(%s): %w", qualifiedTableName, err)
```
Comment (amit-yb) [inline yb-voyager/src/tgtdb/yugabytedb.go:1353]: This doesn't tell for which column the error occurred. Is there any way to find that out?

File: yb-voyager/cmd/importData.go [hunk 10]
```diff
@@ -245,6 +249,20 @@ func importData(importFileTasks []*ImportFileTask) {
...
+	if err != nil {
+		utils.ErrExit("failed to get identity columns from meta db: %s", err)
+	}
+	if !found {
+		tables := importFileTasksToTableNames(importFileTasks)
+		TableToIdentityColumnNames = getGeneratedAlwaysAsIdentityColumnNamesForTables(tables)
+		// saving in metadb for handling restarts
+		metaDB.InsertJsonObject(nil, identityColumnsMetaDBKey, TableToIdentityColumnNames)
+	}
+
```
Comment (amit-yb) [inline yb-voyager/cmd/importData.go:262]: Why can't you move this to the newly introduced `disableGeneratedAlwaysAsIdentityColumns()` function?

**Review by amit-yb — APPROVED** [review]
LGTM.

File: yb-voyager/cmd/importData.go (hunk 10 above)
Comment (sanyamsinghal) [reply yb-voyager/cmd/importData.go:262]: done

File: yb-voyager/src/tgtdb/yugabytedb.go (hunk 9 above)
Comment (sanyamsinghal) [reply yb-voyager/src/tgtdb/yugabytedb.go:1353]: we decided to defer this for now as per discussion

**Pull request merged**